### PR TITLE
fix: [CDS-54798]: YAML side changes for Service hooks

### PR DESCRIPTION
**Pull request by TarunUba1**
## Describe your changes
Yaml changes for hoooks
refer : https://harness.atlassian.net/wiki/spaces/CDP/pages/21307260957/Kubernetes+Service+Hooks

## Testing
 
Screenshot 2023-03-14 at 4 27 16 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44934)

### CODE REVIEW

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/visitor/helpers/hook/HookAttributesVisitorHelper.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+public class HookAttributesVisitorHelper implements ConfigValidator {
+  @Override
+  public void validate(Object object, ValidationVisitor visitor) {
+    // Nothing to validate.
+  }
+
+  @Override
+  public Object createDummyVisitableElement(Object originalElement) {
+    HookAttributes hookAttributes = (HookAttributes) originalElement;
+    return ConfigFileAttributes.builder().store(hookAttributes.getStore()).build();
```
Comment (TarunUba1) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/visitor/helpers/hook/HookAttributesVisitorHelper.java:24]: will change it today

**Review by abosii-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 15 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44934)* status: 15 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java` line 47 at r1](https://reviewable.io/reviews/harness/harness-core/44934#-NQA2QMk1sREDQgZhn-9:-NQA2QMk1sREDQgZhn-A:bidvane) ([raw file](https://github.com/harness/harness-core/blob/8c6547ef189c115e5ee4490dbe6a3a87c87c262e/127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java#L47)):*
> ```Java
> 
>   @NotNull @JsonProperty("type") HookConfigType type;
>   @NotNull @JsonProperty("actions") HookConfigActions actions;
> ```

Shouldn't this be a list?

**Review by TarunUba1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44934)* status: 15 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java` line 47 at r1](https://reviewable.io/reviews/harness/harness-core/44934#-NQA2QMk1sREDQgZhn-9:-NQPB2jA0kSamb4hbiT6:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/8c6547ef189c115e5ee4490dbe6a3a87c87c262e/127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java#L47)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

Shouldn't this be a list?
</blockquote></details>

Done.

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  @JsonProperty(YamlNode.UUID_FIELD_NAME)
+  @Getter(onMethod_ = { @ApiModelProperty(hidden = true) })
+  @ApiModelProperty(hidden = true)
+  private String uuid;
+
+  @NotNull @EntityIdentifier String identifier;
+
+  @NotNull @JsonProperty("actions") List<HookConfigActions> actions;
+
+  @NotNull @JsonProperty("type") StoreConfigType storetype;
```
Comment (abosii-harness) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java:63]: Can you give a try to @Schema(type = "string", allowableValues = {"Account", "Repo"})

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+  @NotNull
+  @JsonTypeInfo(use = NAME, property = "type", include = EXTERNAL_PROPERTY, visible = true)
```
Comment (abosii-harness) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java:66]: Shouldn't this be property = "storeType?"

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  @Wither
+  StoreConfig store;
+
+  // For Visitor Framework Impl
+  @Getter(onMethod_ = { @ApiModelProperty(hidden = true) }) @ApiModelProperty(hidden = true) String metadata;
+
+  @Builder
+  public HookConfig(
```
Comment (abosii-harness) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java:74]: Can be on class

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HooksOutcome.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.data.annotation.TypeAlias;
+
+@OwnedBy(CDP)
+@Value
+@Builder
+@TypeAlias("hooksOutcome")
+@JsonTypeName("hooksOutcome")
+@FieldNameConstants(innerTypeName = "HooksOutcomeKeys")
+@RecasterAlias("io.harness.cdng.hooks.HooksOutcome")
+public class HooksOutcome {
```
Comment (abosii-harness) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HooksOutcome.java:30]: should be single Hook

**Review by abosii-harness — CHANGES_REQUESTED** [review]

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java (hunk 2 above)
Comment (TarunUba1) [reply 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java:63]: Unsucessful, everything roams around string, nothing around to specify obejct. I saw for config files the way they control it is they show in UI only Harness option.

**Review by abosii-harness — COMMENTED** [review]
> Reviewed 3 of 13 files at r2, 7 of 12 files at r5, 6 of 6 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44934)* status: 16 file(s) to review. 1 change requests. 0 of 2 approvals obtained.

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java (hunk 3 above)
Comment (TarunUba1) [reply 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java]: done

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java (hunk 4 above)
Comment (TarunUba1) [reply 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HookConfig.java]: done

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HooksOutcome.java (hunk 5 above)
Comment (TarunUba1) [reply 127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/HooksOutcome.java]: done

**Review by abosii-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44934)* status: 16 file(s) to review. 1 of 2 approvals obtained.

**Review by namantalaycha10 — APPROVED** [review]
> Reviewed 3 of 13 files at r2, 7 of 12 files at r5, 6 of 6 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44934)* status: :shipit: complete! ✓

**Review by abhinavsingh1196 — DISMISSED** [review]
> Reviewed 3 of 13 files at r2, 7 of 12 files at r5, 6 of 6 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44934)* status: :shipit: complete! ✓

___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHookActions.java` line 19 at r7](https://reviewable.io/reviews/harness/harness-core/44934#-NQWC2QO1b_iNJ4xIZ6W:-NQWC2QO1b_iNJ4xIZ6X:bv9if3s) ([raw file](https://github.com/harness/harness-core/blob/f3b3dde89901f918ad52bb21851a2fc2aecea53d/127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHookActions.java#L19)):*
> ```Java
> 
> @OwnedBy(HarnessTeam.CDP)
> public enum ServiceHookActions {
> ```

nit: should be ServiceHookAction? we can make this change in a different PR. I don't want to block for this.

**Comment by yogesh-chauhan** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHookConstants.java` line 10 at r7](https://reviewable.io/reviews/harness/harness-core/44934#-NQWXIkqBt2Msi0K48k5:-NQWXIkqBt2Msi0K48k6:b-mg95um) ([raw file](https://github.com/harness/harness-core/blob/f3b3dde89901f918ad52bb21851a2fc2aecea53d/127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHookConstants.java#L10)):*
> ```Java
> package io.harness.cdng.hooks;
> 
> public interface ServiceHookConstants {
> ```

nit: better to use utility class for storing set of constant strings 
https://stackoverflow.com/questions/29382728/constant-interface-anti-pattern-clarification

**Comment by yogesh-chauhan** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHookOutcome.java` line 30 at r7](https://reviewable.io/reviews/harness/harness-core/44934#-NQWX_jZ-35Zmf1kp9ba:-NQWX_jZ-35Zmf1kp9bb:b3l0wqn) ([raw file](https://github.com/harness/harness-core/blob/f3b3dde89901f918ad52bb21851a2fc2aecea53d/127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHookOutcome.java#L30)):*
> ```Java
> @FieldNameConstants(innerTypeName = "ServiceHookOutcomeKeys")
> @RecasterAlias("io.harness.cdng.hooks.ServiceHookOutcome")
> public class ServiceHookOutcome {
> ```

should it not extend `io.harness.pms.sdk.core.data.Outcome`

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]
> Reviewed 3 of 13 files at r2, 6 of 12 files at r5, 3 of 6 files at r7.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44934)* status: :shipit: complete! ✓

___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHook.java` line 63 at r7](https://reviewable.io/reviews/harness/harness-core/44934#-NQWX0dc2QTgu_B4MCn-:-NQWX0dc2QTgu_B4MCn0:b6xavd3) ([raw file](https://github.com/harness/harness-core/blob/f3b3dde89901f918ad52bb21851a2fc2aecea53d/127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHook.java#L63)):*
> ```Java
>   @NotNull @EntityIdentifier String identifier;
> 
>   @NotNull @JsonProperty("actions") List<ServiceHookActions> actions;
> ```

probably use @NotEmpty here

**Review by TarunUba1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44934)* status: 5 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHook.java` line 63 at r7](https://reviewable.io/reviews/harness/harness-core/44934#-NQWX0dc2QTgu_B4MCn-:-NQWjB5T-Yq4Y53GquJp:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f3b3dde89901f918ad52bb21851a2fc2aecea53d/127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHook.java#L63)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

probably use @NotEmpty here
</blockquote></details>

Done.
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHookActions.java` line 19 at r7](https://reviewable.io/reviews/harness/harness-core/44934#-NQWC2QO1b_iNJ4xIZ6W:-NQWjOXJ1gus4oDkRxoC:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/f3b3dde89901f918ad52bb21851a2fc2aecea53d/127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHookActions.java#L19)):*
<details><summary><i>Previously, abhinavsingh1196 (Abhinav Singh) wrote…</i></summary><blockquote>

nit: should be ServiceHookAction? we can make this change in a different PR. I don't want to block for this.
</blockquote></details>

Done
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHookConstants.java` line 10 at r7](https://reviewable.io/reviews/harness/harness-core/44934#-NQWXIkqBt2Msi0K48k5:-NQWjQCT2W2O2NTqg19N:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/f3b3dde89901f918ad52bb21851a2fc2aecea53d/127-cd-nextgen-entities/src/main/java/io/harness/cdng/hooks/ServiceHookConstants.java#L10)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

nit: better to use utility class for storing set of constant strings 
https://stackoverflow.com/questions/29382728/constant-interface-anti-pattern-clarification
</blockquote></details>

Done

**Review by TarunUba1 — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44934#-:-NQWk03i6Eni_D1suHTR:bnfp4nl)* status: 5 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 12 files at r5, 1 of 6 files at r7, 5 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44934)* status: 5 file(s) to review. 1 of 2 approvals obtained.

**Review by namantalaycha10 — APPROVED** [review]
> Reviewed 5 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44934)* status: :shipit: complete! ✓

**Pull request merged**